### ISSUE TITLE
sops -d fails to open /dev/stdout when running with privileges of another user

### ISSUE THREAD

**Issue by bhourigan** [mozilla/sops#37]
When I assume another users privileges I'm unable to run `sops -d cat.json` and inspect encrypted data, it looks like sops is unable to get write permissions on /dev/stdout.

I can, however, run `sops -d cat.json | cat` to inspect the contents.

Steps to reproduce:
```
bhourigan@system$ sudo su
root@system# su machineaccount
machineaccount@system$ sops -d cat.json
Traceback (most recent call last):
  File "/usr/local/bin/sops", line 9, in <module>
    load_entry_point('sops==1.3', 'console_scripts', 'sops')()
  File "/usr/local/lib/python2.7/dist-packages/sops/__init__.py", line 226, in main
    write_file(tree, path=dest, filetype=otype)
  File "/usr/local/lib/python2.7/dist-packages/sops/__init__.py", line 1034, in write_file
    fd = open(path, "wb")
IOError: [Errno 13] Permission denied: u'/dev/stdout'
```

Path permission/ownership chain:
```
$ ls -ald /dev/stdout
lrwxrwxrwx 1 root root 15 Jan 19 20:10 /dev/stdout -> /proc/self/fd/1
$ ls -ald /proc/self/fd/1
lrwx------ 1 machineaccount machineaccount 64 Jan 22 14:29 /proc/self/fd/1 -> /dev/pts/0
$ ls -ald /dev/pts/0
crw--w---- 1 bhourigan tty 136, 0 Jan 22 14:29 /dev/pts/0
```

Is this a feature, or a bug?

**Comment by jvehent** [comment]
@gdestuynder : any idea what's happening here?

**Comment by gdestuynder** [comment]
this is a common issue when using sudo or su to switch users, since your pts still belong to the original user (the pts is the tty used to control your terminal, which is indeed attached to the user you ssh'd/logge'd in as, and is thus correct), anything that attempts to write to it without the proper rights will fail (most commonly gpg does this too)
this doesn't happen when /dev/null is using a manually managed devfs (old linuxes) in this case.

fixes that i know of include:
- not writing to /dev/null
- having machineaccount part of the tty group
- spawning a new pts as the new user (for ex with the `script', `tmux', `screen' commands or through consolekit/loginctl/your-tty-allocator)

**Comment by bhourigan** [comment]
is it writing to the pty directly because of security concerns? any opposition to just using sys.stdout?

**Comment by gdestuynder** [comment]
the later writes to the already opened stdout (fd 1) while the former attempts to re-open stdout through the filesystem (which gets denied, because of ownership and file modes/permissions)
i'm not sure if there's any direct security concern (in fact, the file system did its job here), albeit there's a bit of room for error if somehow /dev/stdout links elsewhere. Otherwise, there is a slight performance concern though.

Most programs use stdout (alias of "1") or sys.write.stdout() on python (or os.write(1, b"hi") which would also work!).

TLDR:
- not writing to /dev/null (but to fd 1) is the commonly accepted behavior (AFAIK ;-) that fixes this

**Comment by jvehent** [comment]
got it. have a patch for it by simply assigning the fd to `sys.stdout` in write_file()
